compiler/bytecode/statement_generation.go

```diff
@@ -155,25 +157,40 @@ func (g *Generator) compileDefStmt(is *InstructionSet, stmt *ast.DefStatement, s
 	newIS.isType = MethodDef
 
 	for i := 0; i < len(stmt.Parameters); i++ {
-		var argType int
 		switch exp := stmt.Parameters[i].(type) {
 		case *ast.Identifier:
-			argType = NormalArg
 			scope.localTable.setLCL(exp.Value, scope.localTable.depth)
+
+			newIS.argTypes = append(newIS.argTypes, NormalArg)
 		case *ast.AssignExpression:
-			argType = OptionedArg
 			exp.Optioned = 1
 			g.compileAssignExpression(newIS, exp, scope, scope.localTable)
+
+			newIS.argTypes = append(newIS.argTypes, OptionedArg)
 		case *ast.PrefixExpression:
 			if exp.Operator != "*" {
 				continue
 			}
-			argType = SplatArg
 			ident := exp.Right.(*ast.Identifier)
 			scope.localTable.setLCL(ident.Value, scope.localTable.depth)
-		}
 
-		newIS.argTypes = append(newIS.argTypes, argType)
+			newIS.argTypes = append(newIS.argTypes, SplatArg)
+		case *ast.HashExpression:
+			for key, exp := range exp.Data {
+				var argType int
+
+				if exp == nil {
+					argType = RequiredKeywordArg
+				} else {
+					g.compileExpression(newIS, exp, scope, scope.localTable)
+					index, depth := scope.localTable.setLCL(key, scope.localTable.depth)
```

Comment:
I think maybe we shouldn't use current local table here anymore. Because that only operate based on index, but what we need to do might need more information than that. So maybe we should upgrade our local table system, or use a new data structure for the keyword arguments.

Reply:
Yes, I agree that current index based operation already reached its limit. Let me think about how to upgrade local table so that somehow it can handle keyword arguments and clear up `evalMethodObject` method. (Also, I think we can improve argTypes to handle keyword argument, but I'm pretty sure it will end up with messing up `evalMethodObject` method even more....)

Reply:
I'm referring to YARV's local table implementation now

Reply:
Great job, but maybe we should upgrade the local table at another branch (base on current master). Because that'll be another new (and might be large) refactoring I think.

Reply:
Hmm, do you think I should limit this pr to only parsing of keyword arguments? Since without refactoring local table class, I do not think it is possible to implement  code generation part of keyword argument.

Reply:
Nope 
I mean you should implement local table in other PR, and rebase/merge master this once it's merged into master

Reply:
You got it! Let me work on that on another pr! 👍

Reply:
@st0012 Hmm, I'm still stuck in here... Actually I figured out that Ruby use a hidden argument when they call a method with a keyword argument. For example `test` method like below

```
def test(a, b: 1, c: 2); end
```
When Ruby call `test` method like this

```
test(10, b: 20, c: 30)
```

Ruby actually push the hash of defaults as a fourth argument, so now the arguments are those four below.
```
10
b: 20
c: 30
{ b: 1, c: 2 }
```

I tried to do the same thing but I could not figure out the way I could access the default hash from `thread.go`. I probably need to improve typed for that purpose but do you have any better ideas?

Thanks!

Reply:
@shuheiktgw Ok, I'll also start working on this on the weekend.

Reply:
You got it! Thanks!